src/libsnark.ml

```diff
@@ -2701,19 +2930,59 @@ struct
       let v = Fqk.to_elts Fqk.one in
       Mnt6_0.Field.Vector.length v = 4
 
+    module Fqe =
+      Fqe.Make
+        (Mnt6_0.Field.Vector)
+        (Fqe.Bind
+           (Ctypes_foreign)
+           (struct
+             let prefix = with_prefix Mnt4_0.prefix "fqe"
+           end)
+           (Mnt6_0.Field.Vector))
+
     module G2 = struct
-      module T = Group (Mnt4_0.Field) (Mnt4_0.Bigint.R) (Mnt6_0.Field.Vector)
+      module T = Group (Mnt4_0.Field) (Mnt4_0.Bigint.R) (Fqe)
+
+      module Coefficients = struct
+        module T = Group_coefficients (Fqe)
+
+        include T.Make
+                  (T.Bind
+                     (Ctypes_foreign)
+                     (struct
+                       let prefix = with_prefix Mnt4_0.prefix "g2"
+                     end))
+      end
 
       include T.Make
                 (T.Bind
                    (Ctypes_foreign)
                    (struct
                      let prefix = with_prefix Mnt4_0.prefix "g2"
                    end))
-                   (Mnt6_0.Field.Vector)
+                   (Fqe)
+                (struct
+                  let subgroup = `Check_subgroup_with_order Mnt4_0.field_size
+
+                  let on_curve (x, y) =
+                    let open Fqe in
```

Comment:
Might be worth using mutability here, as below?

Reply:
It requires adding mutable operators for the field extension multiplication and addition to libff. Addition is easy to do but multiplication is complicated because the algorithm requires additional storage space. Might still be possible to save some allocations though.

Reply:
Okay. If we notice it being a problem later, we can come back and do this.